Support for having LighthouseJobs and Tekton jobs in multiple namespaces

keskad opened the pull request:
Part of https://github.com/jenkins-x/lighthouse/pull/1424

More context and related changes are there: https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/425 and in main PR in lighthouse (https://github.com/jenkins-x/lighthouse/pull/1424)

All changes there are done in context of parent PR (namespaces support, multitenacy)

Release notes
--------------------

### Little steps to parametrize namespace selection in all JX places

- `jx gitops webhook update` have additional commandline switch `--dev-ns` to be able to switch to other namespace (a little step for making Jenkins X not aligned to only one namespace)

### Cluster scoped pods clean up to support running jobs in multiple namespaces

- `jx gitops gc pods` is able to search for Jenkins X namespaces labelled by bootjob with the environment name and then execute pods clean up only in those namespaces (required `RoleBinding` per namespace - this is done in separate repo where helms are - define a list of namespaces in `jxJobsNamespaces` in your `jx-global-values.yaml` to propagate bindings)

### Fix that enables multi-namespaced Helm releases inside bootrepo helmfiles

- `jx gitops namespace --dir-mode --dir config-root/namespaces` is fixed to not enforce namespace when `.metadata.namespace` is already defined in yaml file. This potentially broken some helms that had cross-namespace resources such as RoleBindings. Instead of changing namespace in a resource, the resource is moved to directory named same as `.metadata.namespace`

- `jx gitops helmfile move` is fixed to not enforce namespace when `.metadata.namespace` is already defined in yaml file. This potentially broken some helms that had cross-namespace resources such as RoleBindings.

keskad commented:
Added a linked PR: https://github.com/jenkins-x/jx-helpers/pull/362

keskad commented:
Added a linked PR: https://github.com/jenkins-x/jx-helpers/pull/362

keskad commented on go.mod:
```diff
@@ -20,6 +20,7 @@ require (
 	github.com/jenkins-x/jx-kube-client/v3 v3.0.2
 	github.com/jenkins-x/jx-logging/v3 v3.0.6
 	github.com/jenkins-x/lighthouse-client v0.0.295
+	github.com/keskad/jx-helpers/v3 v3.0.67-0.20220312071418-f6abb1196975  // todo: revert before merging and bump to latest version
```
Before merging please merge https://github.com/jenkins-x/jx-helpers/pull/362 , then switch there back to JX repository and bump a version.

keskad commented on pkg/cmd/namespace/namespace.go:
```diff
@@ -5,24 +5,28 @@ import (
 	"github.com/jenkins-x/jx-helpers/v3/pkg/termcolor"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/yamls"
 	"github.com/jenkins-x/jx-logging/v3/pkg/log"
+	"github.com/keskad/jx-helpers/v3/pkg/kyamls" // todo: revert before merging
```
Before merging please merge https://github.com/jenkins-x/jx-helpers/pull/362 , then switch there back to JX repository and bump a version.

keskad commented on pkg/cmd/namespace/namespace_test.go:
```diff
@@ -1,16 +1,16 @@
-	"github.com/jenkins-x-plugins/jx-gitops/pkg/cmd/namespace"
 	"github.com/jenkins-x/jx-helpers/v3/pkg/files"
-	"github.com/jenkins-x/jx-helpers/v3/pkg/kyamls"
+	"github.com/keskad/jx-helpers/v3/pkg/kyamls" // todo: revert before merging
```
Before merging please merge https://github.com/jenkins-x/jx-helpers/pull/362 , then switch there back to JX repository and bump a version.

keskad commented:
Possibly related
---------------------

The fixes to `jx gitops namespace` and `jx helmfile move` are possibly fixing https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/412

I noticed that it resolved an issue with `gc-jobs` job not being created in proper namespace, so the bootjobs `kind: Job` were not cleaned up.

The resource is defined following:

```
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: jx-gcjobs
  namespace: {{ .Values.gc.jobs.cronjob.ns | default "jx-git-operator" | quote }}
```

And because of the bug it was always placed in `jx` instead of `jx-git-operator`, now I can see that it was fixed by the way:

```
 7 files changed, 27 insertions(+), 5 deletions(-)
 create mode 100644 config-root/cluster/namespaces/jx-git-operator.yaml
 create mode 100644 config-root/cluster/namespaces/jx-test1.yaml
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/gcjobs-rb.yaml (93%)
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/gcjobs-role.yaml (92%)
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/jx-gcjobs-cronjob.yaml (97%)
 rename config-root/namespaces/{jx => jx-git-operator}/jxboot-helmfile-resources/jx-gcjobs-sa.yaml (89%)
 rename config-root/namespaces/{jx => jx-test1}/jxboot-helmfile-resources/gcpods-jx-test1-rb.yaml (96%)
```

So the `gc-jobs` was placed where it was planned to.

keskad commented:
Ready for review :smile: 
@yelhouti, @ankitm123? :slightly_smiling_face:

ankitm123 commented on pkg/cmd/helmfile/move/move_test.go:
```diff
@@ -1,12 +1,12 @@
-package move_test
+package move
```
why this change? We want to only test exported functions.

keskad replied:
Fixed.

ankitm123 commented on pkg/cmd/namespace/namespace_test.go:
```diff
@@ -1,14 +1,14 @@
-package namespace_test
+package namespace
```
same, keep it as namespace_test.

keskad replied:
Sure.

ankitm123 reviewed (commented):
took a quick look, will review more tonight :)

ankitm123 commented on pkg/cmd/gc/pods/pods.go:
```diff
@@ -121,3 +157,28 @@ func (o *Options) MatchesPod(pod *corev1.Pod) (bool, time.Duration) {
+	if err != nil {
+		return []string{}, errors.Wrap(err, "cannot list namespaces to look for pods")
+	}
+	for _, n := range namespaces.Items { //nolint:gocritic
```
why nolint? Does it give false positive? or is it because of `_, n` instead of using `for k := range ...`

keskad replied:
Fixed.

ankitm123 commented on pkg/cmd/gc/pods/pods.go:
```diff
@@ -121,3 +157,28 @@ func (o *Options) MatchesPod(pod *corev1.Pod) (bool, time.Duration) {
+	if err != nil {
+		return []string{}, errors.Wrap(err, "cannot list namespaces to look for pods")
+	}
+	for _, n := range namespaces.Items { //nolint:gocritic
```
why nolint? Does it give false positive? or is it because of `_, n` instead of using `for k := range ...`

keskad replied:
Do you have a suggestion? It says loop iteration = ~400 bytes and I should use a pointer I don't have an idea how to do this, Kubernetes librarry does not return a pointer.

ankitm123 replied:
```
for k := range namespaces.Items {
  n := namespaces.Items[k]
  labels := n.GetLabels()
```

keskad replied:
Thanks, will try it :grin:

keskad replied:
Thanks, fixed.

ankitm123 commented on pkg/cmd/gc/pods/pods_test.go:
```diff
@@ -0,0 +1,61 @@
+package pods
```
`pods_test` for package name.

keskad replied:
Same as https://github.com/jenkins-x-plugins/jx-gitops/pull/835#discussion_r826644017

keskad replied:
Fixed.

ankitm123 commented on pkg/cmd/gc/pods/pods_test.go:
```diff
@@ -0,0 +1,61 @@
+package pods
```
`pods_test` for package name.

keskad replied:
I would not be able to test internal methods if I do so. What do you think?

ankitm123 replied:
we should not be testing internal methods, only test exported functions (there are exceptions to the rules though). This will force us to think about the external API.

keskad replied:
Hmm... but then tests will become more functional tests (testing more facades than single units of code, less precise) than unit testing. But if it is ok for you then I can change.

keskad replied:
Ok, I will try to test just exported methods.

ankitm123 commented on pkg/cmd/namespace/namespace.go:
```diff
@@ -181,25 +191,105 @@ func (o *Options) lazyCreateNamespaceResource(ns string) error {
 // UpdateNamespaceInYamlFiles updates the namespace in yaml files
-func UpdateNamespaceInYamlFiles(dir string, ns string, filter kyamls.Filter) error { //nolint:gocritic
-	modifyFn := func(node *yaml.RNode, path string) (bool, error) {
+func UpdateNamespaceInYamlFiles(rootDir string, dir string, ns string, filter kyamls.Filter, shouldMoveFiles bool) ([]string, error) { //nolint:gocritic
```
`rootDir, dir, ns string` should fix the gocritic linter warning i think 🤔

keskad replied:
Yeah, I guess my IDE appended that nolint thing :thinking:

ankitm123 commented on pkg/cmd/namespace/namespace.go:
```diff
@@ -181,25 +191,105 @@ func (o *Options) lazyCreateNamespaceResource(ns string) error {
 // UpdateNamespaceInYamlFiles updates the namespace in yaml files
-func UpdateNamespaceInYamlFiles(dir string, ns string, filter kyamls.Filter) error { //nolint:gocritic
-	modifyFn := func(node *yaml.RNode, path string) (bool, error) {
+func UpdateNamespaceInYamlFiles(rootDir string, dir string, ns string, filter kyamls.Filter, shouldMoveFiles bool) ([]string, error) { //nolint:gocritic
```
`rootDir, dir, ns string` should fix the gocritic linter warning i think 🤔

ankitm123 commented on pkg/cmd/gc/activities/activities.go:
```diff
@@ -108,6 +108,8 @@ func NewCmdGCActivities() (*cobra.Command, *Options) {
 	cmd.Flags().DurationVarP(&o.ReleaseAgeLimit, "release-age", "r", time.Hour*24*30, "Maximum age to keep PipelineActivities for Releases")
 	cmd.Flags().DurationVarP(&o.PipelineRunAgeLimit, "pipelinerun-age", "", time.Hour*12, "Maximum age to keep completed PipelineRuns for all pipelines")
 	cmd.Flags().DurationVarP(&o.ProwJobAgeLimit, "prowjob-age", "", time.Hour*24*7, "Maximum age to keep completed ProwJobs for all pipelines")
+	cmd.Flags().BoolVarP(&o.AllNamespaces, "all-namespaces", "", false, "Work in all namespaces (requires a ClusterRole)")
```
need to run `make docs` to update the documentation.

keskad replied:
Sure, will do

ankitm123 commented on pkg/cmd/gc/activities/activities.go:
```diff
@@ -108,6 +108,8 @@ func NewCmdGCActivities() (*cobra.Command, *Options) {
 	cmd.Flags().DurationVarP(&o.ReleaseAgeLimit, "release-age", "r", time.Hour*24*30, "Maximum age to keep PipelineActivities for Releases")
 	cmd.Flags().DurationVarP(&o.PipelineRunAgeLimit, "pipelinerun-age", "", time.Hour*12, "Maximum age to keep completed PipelineRuns for all pipelines")
 	cmd.Flags().DurationVarP(&o.ProwJobAgeLimit, "prowjob-age", "", time.Hour*24*7, "Maximum age to keep completed ProwJobs for all pipelines")
+	cmd.Flags().BoolVarP(&o.AllNamespaces, "all-namespaces", "", false, "Work in all namespaces (requires a ClusterRole)")
```
need to run `make docs` to update the documentation.

keskad replied:
Sure will do

ankitm123 commented on pkg/cmd/helmfile/move/move.go:
```diff
@@ -311,3 +315,21 @@ func (o *Options) moveFilesToClusterOrNamespacesFolder(dir, ns, releaseName, cha
+	return filepath.Join(o.NamespacesDir, GetTargetNamespace(node, path, ns), pathName)
+}
+
+func GetTargetNamespace(node *yaml.RNode, path, ns string) string {
```
Will this be used outside of jx-gitops? If not we dont need to export these, unless u have it this way to add tests for these functions ...

keskad replied:
I didn't wanted to export it, but you requrested to set package name to `move_test` and I was unable to test it if it was not exported.

keskad replied:
Fixed.

ankitm123 commented on pkg/cmd/helmfile/move/move.go:
```diff
@@ -311,3 +315,21 @@ func (o *Options) moveFilesToClusterOrNamespacesFolder(dir, ns, releaseName, cha
+	return filepath.Join(o.NamespacesDir, GetTargetNamespace(node, path, ns), pathName)
+}
+
+func GetTargetNamespace(node *yaml.RNode, path, ns string) string {
```
Will this be used outside of jx-gitops? If not we dont need to export these, unless u have it this way to add tests for these functions ...

keskad replied:
Same here as https://github.com/jenkins-x-plugins/jx-gitops/pull/835#discussion_r826643000

keskad replied:
I don't think it would be used outside. Fixed.

keskad commented:
I just commited fixes, @ankitm123 :slightly_smiling_face:

ankitm123 commented on pkg/cmd/webhook/update/update.go:
```diff
@@ -88,6 +89,7 @@ func NewCmdWebHookVerify() (*cobra.Command, *Options) {
 	cmd.Flags().StringVarP(&o.HMAC, "hmac", "", "", "Don't use the HMAC token from the cluster, use the provided token")
 	cmd.Flags().StringVarP(&o.Endpoint, "endpoint", "", "", "Don't use the endpoint from the cluster, use the provided endpoint")
 	cmd.Flags().BoolVarP(&o.WarnOnFail, "warn-on-fail", "", false, "If enabled lets just log a warning that we could not update the webhook")
+	cmd.Flags().StringVarP(&o.EnforceNamespace, "enforce-namespace", "", "", "Disable namespace auto detection, instead use specified namespace")
```
can be confusing, how about &DevNamespace? 
`StringVarP(&o.DevNamespace, "dev-ns",`
This argument is ultimately setting/getting the dev namespace. wdyt?

keskad replied:
Yup. The autodetection mechanism is taking the namespace from the `team` label of the namespace as I remember. It seems also that the concept of `team` was in JX v2.

keskad replied:
Will change to dev-ns.

ankitm123 reviewed (commented):
minor comments, overall looks great!

keskad commented on pkg/cmd/gc/pods/pods.go:
```diff
@@ -121,3 +174,29 @@ func (o *Options) MatchesPod(pod *corev1.Pod) (bool, time.Duration) {
+			}
+		}
+
+		if n.Name == "jx-git-operator" {
```
Sure.

keskad commented on pkg/cmd/gc/pods/pods.go:
```diff
@@ -121,3 +174,29 @@ func (o *Options) MatchesPod(pod *corev1.Pod) (bool, time.Duration) {
 	return age > o.Age, age
 }
+
+// findJXNamespaces looks for namespace names that are owned by Jenkins X
```
Happy to read :)

keskad commented on pkg/cmd/gc/pods/pods.go:
```diff
@@ -51,54 +58,100 @@ func NewCmdGCPods() (*cobra.Command, *Options) {
+		// run in all namespaces that are owned by Jenkins X
+		namespaces, err = o.findJXNamespaces(ctx)
+		if err != nil {
+			return []string{}, errors.Wrap(err, "cannot perform garbage collection of pods")
```
Sure

ankitm123 commented on docs/cmd/jx-gitops_webhook_update.md:
````diff
@@ -28,6 +28,7 @@ Updates the webhooks for all the source repositories optionally filtering by own
 ```
   -b, --batch-mode                 Runs in batch mode without prompting for user input
       --endpoint string            Don't use the endpoint from the cluster, use the provided endpoint
+      --enforce-namespace string   Disable namespace auto detection, instead use specified namespace
````
Need to re-run `make docs`

keskad replied:
Ups :grin:

keskad replied:
Done

keskad commented:
Fixed code review issues, rebased to properly name commits so the chanelog would be nice, added release changelog.

One thing - I accidently touched msvticket's commit description, and it have changed commit id now, sorry for that, don't know how to revert that.

ankitm123 commented:
> Fixed code review issues, rebased to properly name commits so the chanelog would be nice, added release changelog.
> 
> One thing - I accidently touched msvticket's commit description, and it have changed commit id now, sorry for that, don't know how to revert that.

use git reflog, may be?

keskad commented:
> > Fixed code review issues, rebased to properly name commits so the chanelog would be nice, added release changelog.
> > One thing - I accidently touched msvticket's commit description, and it have changed commit id now, sorry for that, don't know how to revert that.
> 
> use git reflog, may be?

OK, I made a new local branch from original repo, cherry-picked my commits and pushed there.

keskad commented:
Possibly related: https://github.com/jenkins-x-plugins/jx-gitops/pull/514 (may be closed?)

ankitm123 commented:
/ok-to-test

nicolaerosia commented:
does this fix https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/412 ?

yelhouti commented:
Yes it is supposed to

keskad commented:
@nicolaerosia Yes, but also linked PR is needed to resolve it: https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/425/files#diff-6f5f287a8a9a6c3fd8cfd6eaa22fb0ea4b466240f2bcc6f507d233d11ce90df8R13

Without that change in RoleBinding it will get access denied still. So just a little moment more is needed in the review process I think :grin:

ankitm123 commented:
/lgtm
